lib/posix/pthread_cond.c

```diff
@@ -52,20 +52,30 @@ static int cond_wait(pthread_cond_t *cv, pthread_mutex_t *mut,
 
 int pthread_cond_signal(pthread_cond_t *cv)
 {
+	struct k_thread *thread;
 	k_spinlock_key_t key = k_spin_lock(&z_pthread_spinlock);
 
-	_ready_one_thread(&cv->wait_q);
+	thread = _ready_one_thread(&cv->wait_q);
+	if (thread != NULL) {
+		arch_thread_return_value_set(thread, 0);
```

Comment:
Actually on further thought: I withdraw my lack of objections.  This actually won't work. :)

At the point where _ready_one_thread() has returned, the thread has already been readied and **may already have run** (on an SMP system with an idle core, it's basically guaranteed to be running already).  It's too late to set the return value here.  My guess is that this resolves the error seen simply because the thread running this code is higher priority.

Reply:
I assumed that because the POSIX lick was held it was safe

Reply:
If you're in a nested spinlock, then yes: you're protected from preemption on uniprocessor systems.  But on SMP, ready means ready and the scheduler is free to run that on any core available.